### ISSUE TITLE
No debug file generated if fatal error occurred 

### ISSUE THREAD

**Issue by Soukyuu** [uncrustify/uncrustify#1045]
[//]: # "  If the issue is connected to existing Uncrustify options please, if possible, add the"
[//]: # "  following information to ease up the process:"
[//]: # "  • a link to a debug file:"
[//]: # "      generated with: 'uncrustify -p debug.txt -c pathToUsedConfig.cfg -f toBeFormatedFile.cpp' "
[//]: # "      Example: [debug.txt](https://linkToTheFile)"
[//]: # "      Example hosters for debug files: pastbin.com, gist.github.com, ..."
[//]: # "      The used config file is included in the debug file and does not need to be included here."
[//]: # 
[//]: # "  • include a small but complete test file that will be uncrustifyed"
[//]: # "  • include the generated results"
[//]: # "  • include the expected results"
[//]: # 
[//]: # "  ✋ please add a line containing ``` above and below of each of those three code sections"
[//]: # 
[//]: # "  • include the current version of your Uncrustify executable"
[//]: # "      printout via 'uncrustify -v'"
[//]: # "      Example: current version: uncrustify 0.63"
[//]: # "      or if possible additionally with the git sha of the commit"
[//]: # "                    current version: uncrustify 0.63 dc7b412"
[//]: # 
[//]: # "  • if possible include a version that worked"
[//]: # "      Example: working version: uncrustify 0.63"
[//]: # "      or"
[//]: # "                    working version: uncrustify 0.63 2a5e88f"
With uncrustify 0.64 c5c95d2, compiled with MSVC2013 x64 (replacing `%zu` with `%Iu` so that it doesn't crash), there is no debug file generated by the `-p` option if uncrustify encounters a fatal error, i.e. `unmatched BRACE_CLOSE`.

Fixing the file manually and running the same command results in a debug.txt generated. The behavior is problematic, because I believe I have a file where uncrustify does not correctly detect the braces, resulting in that error to be thrown even though the file is valid (it compiles and runs correctly).

**Comment by guy-maurel** [comment]
same answer as #1048

**Comment by Soukyuu** [comment]
I believe the debug output object itself is intact, but is not being flushed to disk (haven't looked at the code, though). I think this can be solved by also flushing the debug output to disk when a fatal error occurs.

**Comment by guy-maurel** [comment]
Can you make a proposal? I don't know how.

**Comment by Soukyuu** [comment]
From poking around the code I see you just do a `printf(stderr, ...) ` followed by `exit(EXIT_FAILURE);` on a fatal error. I see you have a `log_flush(bool)` function that seems to be writing the buffer to stream. It is called from within the `log_end()` function, which in turn is part of any logging function. However, `log_flush(bool)` is only called after the buffer has reached a certain size.

Couldn't you use `LOG_FMT(...);` to write the error to the log file, followed by a call to `log_flush(true);`, __then__ exit the program? That should flush the buffer to disk instead of just sending it to digital nirvana.

It would probably require moving the declaration of `log_flush()` to logger.h to make it visible. Or maybe introduce a `log_cleanup()` function that does the job if making `log_flush()` visible is undesired.

That's provided `log_flush()` does what I think it does....

**Comment by guy-maurel** [comment]
Please test the #1188

**Comment by guy-maurel** [comment]
Please have a look to:
https://github.com/uncrustify/uncrustify/wiki/Known-Problems

**Comment by gdamore** [comment]
This is really bad.  While handling the C preprocessor generically is likely problematic, it would
be really really good if we could at least handle extern "C" { constructs... these are super common place and really required for any project that wants to expose C header files to C++. 

The current behavior represents a regression -- and means I can't really use uncrustify 0.65. :(

**Comment by CDanU** [comment]
Yes, we are still discussing what to do about this: see #1196, #1228

**Comment by mihaipopescu** [comment]
Yes, wanted to say the same thing. Seems like it's worse than I thought. Let's try find a solution quickly.

**Comment by gdamore** [comment]
This, combined with the instability of uncrustify in general (by
instability I mean that different versions of uncrustify generate different
output, even given the same config file!) means that I’ve had to move away
from uncrustify . The whole extern “C” thing was kind of my final straw;
I’m settling for the less featureful (but hopefully more reliable)
clang-format for now.

**Comment by mihaipopescu** [comment]
@gdamore We have considered `clang-format` ourselves more than one time exactly on the reasons you've mentioned. We have experienced a lot of regressions ourselves in the past fact that made us stop merging and we diverged a lot. We've recently started to put more work in Uncrustify and fix high priority issues but it's not guarantee that things won't break in the future if there are no tests to catch regressions. We thought `clang-format` had a better lexer and much more mature code base but we've come back to uncrustify for different reasons:

- `clang-format` doesn't support C# which for us is a major setback
- Uncrustify's source code is much simpler and more hackable than on what `clang-format` is build upon (namely libclang tooling).
- `clang-format`, as you've mentioned, has fewer options and features, much geared to suite other big companies needs like Google or Mozilla
- They are reluctant in adding support for new languages. See reaction when they've added javascript: http://comments.gmane.org/gmane.comp.compilers.clang.devel/32785
    > This sounds pretty interesting to me.  The only concern I would have is that we don't want Clang's lexer to have to become a "Grand unified lexer for all languages".  The token rules of languages in these families are related, but different.  Would it be better to have clang-format support using multiple different lexers?
- and:
    > The goal of the clang-format project is more on the side of supporting a limited set of styles really well as opposed to supporting every single style used by a codebase somewhere in the wild.
- On the other hand, uncrustify has a much more opened fervent community of active developers. This can be good and bad but it's only in our powers to make it better.

**Comment by Soukyuu** [comment]
I am not sure why we're discussing issue #1048 in here, and why this one was closed with a link that states the problem described in #1048 is a known problem.

The actual issue reported here is that the debug file is not created on a fatal error, which is still the case even after pull request #1188 was merged. This makes debugging fatal errors impossible,

**Comment by mihaipopescu** [comment]
As we have sidetracked we're not still discussing this issue.

Apparently, this issue was closed because it's a "Known Issue" which was wrong. While I understand why we can't do indentation of that code I don't agree in closing issues that are crashes or false-positive error reporting. And I don't agree in closing issues in general unless the issue has a resolution of some sort.

See also some related issues: #1196, #1228 and probably more recent #1279, #1280.
